### DISCUSSION TITLE
[Axmol 3.x d3d11 implemented] ...  What about directx implemention? what about using SDL and bgfx ?

### DISCUSSION THREAD

**Discussion by meiry** [axmolengine/axmol#790] [General]
Hey 
Any planes to Implement directx for windows ?
Or replace the render layer with something like bgfx ? and maybe the windowing with SDL2 
Now that its only on a voluntary basis you will have hard time to maintain diffrente parts of the engine

**Comment by aismann** [comment]
see here:
Render engine collection #67
Can be closed?

**Comment by meiry** [comment]
Thanks , ok so i saw some half baked and some expiriments , 
But can you please tell me who is the leader of this project ? who decides what will get in in this fork?

**Comment by DelinWorks** [comment]
It's @halx99 he is the leader of this repo

**Comment by DelinWorks** [comment]
> Hey Any planes to Implement directx for windows ?

I don't know if you know already but there is a meh translation layer called `Google Angle` that's good to some extent! you can enable it with `AX_USE_COMPAT_GL`

**Comment by meiry** [comment]
its not optimized , its like you said translation layer , im talking to more direct approach

**Comment by DelinWorks** [comment]
> its not optimized , its like you said translation layer , im talking to more direct approach

yeah I get what you mean.. it sucks tbh because it uses an opengl es mobile backend which is counter-intuitive for performance, also shaders don't work right away and sometimes it crashes on my GPU for no reason and performance is not that great at all

and spending time to learn the directx api just for windows is not a great time investment, bgfx is a great solution

**Comment by DelinWorks** [comment]
This one doesn't have a directX support https://github.com/google/filament

and these two
https://github.com/OGRECave/ogre-next
https://github.com/DiligentGraphics/DiligentEngine

rendering repos seem to focus on 3D and Physically based rendering which this engine isn't intended for unless you guys want it to be a 2D/3D engine combo ;)

but this https://github.com/bkaradzic/bgfx seems like it wraps OpenGL, DirectX, Vulkan and Metal which is what we are looking for, with a bonus of 3D shaders like deferred lighting and basic shadows, toon shading and whatnot, also I used this library before and the API seems to be robust

I think deferred lighting is cool! just look at how many lights you can throw in a scene without pigging the GPU
![image](https://user-images.githubusercontent.com/45469625/183387975-386ec204-64e2-4f35-bca1-5b4c8cd076cd.png)

**Comment by aismann** [comment]
> but this https://github.com/bkaradzic/bgfx seems like it wraps OpenGL, DirectX, Vulkan and Metal which is what we are looking for, with a bonus of 3D shaders like deferred lighting and basic shadows, toon shading and whatnot, also I used this library before and the API seems to be robust

I think also BGFX is the right one.

**Comment by meiry** [comment]
The problem is that this implemention to be production ready on all combos(pc/mobile/web/mac/linux)  is a lot of work . 
Its a shame to take production level engine and abandon it . it will unfortunately toy project . 
i know this engine for 10 years .  
Any good luck to all

**Comment by aismann** [comment]
@meiry 
What are you want?
In your first post you ask about SDL/bgfx and native DirectX support.
We be happy to get your PRs.

**Comment by aismann** [comment]
@halx99 
Please close it.
Seems to be there was a troll here.

**Comment by meiry** [comment]
no sorry , not troll at all , i just wanted to know the state of this project ,i got tthe general idea . 
Its important to others to know also.

**Comment by aismann** [comment]
@meiry 
Axys is on Beta stage.
Please look also on the readme.

**Comment by aismann** [comment]
@halx99 
Please close it.

**Comment by DelinWorks** [comment]
@meiry who said it's gonna be abandoned? the implementation is great because it wraps all the rendering APIs in simple commands, and it shouldn't be hard to implement, also this project is not a toy project and it's a professional production-ready project just like the tool to be implemented.

**Comment by DelinWorks** [comment]
also why do you want d3d11 support? not like this engine is a 3D realistic game engine like unreal engine, if anything we could just switch to vulkan just like godot did, and even then it wouldn't help because vulkan is just like opengl in it's support nature, never gave a shit about that closed api called directx tbh, the time investment to learn and implement it for a single platform is an EXTREME time waste. that's why godot didn't want directx and stuck to vulkan even though they are production ready and professional! they didnt want directx because of the platform dependant bugs that would emerge and the many shaders that have to be written

**Comment by DelinWorks** [comment]
@meiry i see what you mean and i don't consider you a troll in my book :) it's actually a great question!

**Comment by DelinWorks** [comment]
So i've been thinking about it and we can have single support for Vulkan and that's it, as for IOS devices we could use [MoltenVK](https://github.com/KhronosGroup/MoltenVK)
OR we could use BGFX! (which provides metal support out of the box)
Thanks to apple now we'd need to rewrite the rendering api twice if we're going to implement vulkan and translate it to their stupid metal API.. or it can be as simple as including a header file with MoltenVK who knows

**Comment by meiry** [comment]
@DelinWorks  , 
For you questions : Directx is very important for desktop and xbox .  OpenGL in the end is not native on windows OS . 
The Directx question was rasied over and over again in the cocos2d-x days . but never got the time to implement . 
Now for teachnical side . i think that time wise bgfx is the shortest way to get them all . 
But i dont know if its optimal for games .

**Comment by DelinWorks** [comment]
it is optimal! commercial companies built their engines on top of it.
also to make games for Xbox you'd need UWP platform support in the first place and then you'd need a DirectX implementation.

how about on the next release we branch off to `axys v2.0.0` into another branch and any pull requests that try to implement BGFX should pull to the `dev-2.0.0` branch. so that version 1.x.x is already working and maintainable just like Godot does with their project. or we can do the opposite where we work on BGFX on branch `dev` and before doing that we create a branch called `dev-1.x`
@halx99 do you think that's gonna work?

**Comment by halx99** [comment]
it's ok

**Comment by meiry** [comment]
@DelinWorks , 
What commercial companies  build there games around bxfg except wha in the bxfg site ?

**Comment by DelinWorks** [comment]
minecraft. that's what i remember.
also, What do you underestimate about this tool? it works great it supplies a great interface to all APIs... not like this project is gonna be more profissional.

**Comment by meiry** [comment]
i do not underestimate  at all just asking questions

**Comment by aismann** [comment]
Should be close.
All answers can be found on readme and included links.

**Comment by DelinWorks** [comment]
> i do not underestimate at all just asking questions

or we can implement Vulkan only and then translate it using ANGLE for Xbox. I think that's a solution.

**Comment by aismann** [comment]
@DelinWorks, 
We have a roadmap. 
@meiry Look at roadmap plz.

**Comment by DelinWorks** [comment]
Do you mean milestones?

**Comment by aismann** [comment]
> Do you mean milestone.
Yes.

**Comment by DelinWorks** [comment]
so what in there?

**Comment by DelinWorks** [comment]
ohh you mean the `1.1` milestone. I get it we should close this and mark it in there

**Comment by rh101** [comment]
This should be closed, because it's not an issue or a problem with the engine, it's more of a discussion, should be moved to the [Discussions](https://github.com/axys1/axys/discussions) section of this project.

**Reply by halx99** [reply]
converted

**Reply by rh101** [reply]
Thank you

**Comment by aismann** [comment]
Thank you

**Comment by halx99** [comment]
https://github.com/DiligentGraphics/DiligentEngine seems more powerfull, it provide commom graphics APIs to access driver APIs(DirectX, OpenGL, Vulkan, Metal, GLES) directly.

**Reply by halx99** [reply]
![image](https://user-images.githubusercontent.com/6977319/230709379-cfa8584a-1e7a-42cb-aeb3-bcf9bcf5ec6b.png)

**Reply by aismann** [reply]
https://github.com/DiligentGraphics/DiligentEngine
=> Metal is: 2 Available under commercial license - please contact us for details.

![image](https://user-images.githubusercontent.com/8652787/230711473-4196b457-6301-4475-93b7-5a25682c9328.png)

**Reply by DelinWorks** [reply]
the metal api needs a paid license like @aismann said, i think it wouldn't work out.

**Comment by DelinWorks** [comment]
Maybe we can rewrite axmol's renderer in `Vulkan` for all platforms and use `MoltenVK` for MacOS, iOS, and TvOS.. that way we don't have to lean on any abstraction layer that may become obsolete and just use a semi-xplatform api like vulkan. this option seems viable 👍 

Only drawback is that UWP/XBOX apps don't support Vulkan and need a translation layer from vulkan to d3d11 which I don't think that exists.

**Comment by DelinWorks** [comment]
I've looked into vulkan and made a working triangle rendering sample and I partially know how the api works.. except for vertex/index buffers and staging buffers and descriptor pools/sets and texture mapping which are quite extensive topics, they might take some time,  [(Here's a good read to learn vulkan quickly if you're interested)](https://vulkan-tutorial.com/Introduction) I might try getting a working test case using vulkan with glfw to render sprites in axmol and then I'll move to piplines, renderpass, and command buffer implementations into the engine, that way we can have a more efficient and well-supported api for rendering and we can branch out to sophisticated 3D rendering and instancing. (for Apple products as mentioned above MoltenVK is an excellent translation layer maintained by khronos group themselves, the ones that made vulkan, and for XBOX we can fallback to OpenGL support with ANGLE translation to d3d11) you can choose which API to use before creating a GLView and it will select between opengl and vulkan to support backwards compatibility.

**Comment by paulocoutinhox** [comment]
Some work with bgfx was made here:
https://github.com/quicklyslow/cocos2d-bgfx

**Comment by kiranb47** [comment]
https://github.com/facebook/igl

Seems like a good graphics api

**Comment by halx99** [comment]
d3d11 was implemented in branch `dev` now
